go/chat/push.go

```diff
@@ -6,292 +6,483 @@ import (
 )
 
+type messageWaiterEntry struct {
+	vers chat1.InboxVers
+	cb   chan struct{}
+}
+
+type gregorMessageOrderer struct {
+	globals.Contextified
+	utils.DebugLabeler
+	sync.Mutex
+
+	clock   clockwork.Clock
+	waiters map[string][]messageWaiterEntry
+}
+
+func newGregorMessageOrderer(g *globals.Context) *gregorMessageOrderer {
+	return &gregorMessageOrderer{
+		Contextified: globals.NewContextified(g),
+		DebugLabeler: utils.NewDebugLabeler(g, "gregorMessageOrderer", false),
+		waiters:      make(map[string][]messageWaiterEntry),
+		clock:        clockwork.NewRealClock(),
+	}
+}
+
+func (g *gregorMessageOrderer) msgKey(uid gregor1.UID, vers chat1.InboxVers) string {
+	return fmt.Sprintf("%s:%d", uid, vers)
+}
+
+func (g *gregorMessageOrderer) isUIDKey(key string, uid gregor1.UID) bool {
+	toks := strings.Split(key, ":")
+	return toks[0] == uid.String()
+}
+
+func (g *gregorMessageOrderer) latestInboxVersion(ctx context.Context, uid gregor1.UID) (chat1.InboxVers, error) {
+	ibox := storage.NewInbox(g.G(), uid)
+	vers, err := ibox.Version(ctx)
+	if err != nil {
+		return 0, err
+	}
+	return vers, nil
+}
+
+func (g *gregorMessageOrderer) addToWaitersLocked(ctx context.Context, uid gregor1.UID, storedVers,
+	msgVers chat1.InboxVers) (res []messageWaiterEntry) {
+	for i := storedVers + 1; i < msgVers; i++ {
+		entry := messageWaiterEntry{
+			vers: msgVers,
+			cb:   make(chan struct{}),
+		}
+		res = append(res, entry)
+		key := g.msgKey(uid, i)
+		g.waiters[key] = append(g.waiters[key], entry)
+	}
+	return res
+}
+
+func (g *gregorMessageOrderer) waitOnWaiters(ctx context.Context, vers chat1.InboxVers,
+	waiters []messageWaiterEntry) (res chan struct{}) {
+	res = make(chan struct{})
+	go func() {
+		for _, w := range waiters {
+			select {
+			case <-w.cb:
+			case <-ctx.Done():
+				g.Debug(ctx, "waitOnWaiters: context cancelled: waiter: %d target: %d", vers, w.vers)
+			}
+		}
+		close(res)
+	}()
+	return res
+}
+
+func (g *gregorMessageOrderer) cleanupAfterTimeoutLocked(uid gregor1.UID, vers chat1.InboxVers) {
+	for k, v := range g.waiters {
+		if g.isUIDKey(k, uid) {
+			var newv []messageWaiterEntry
+			for _, w := range v {
+				if w.vers != vers {
+					newv = append(newv, w)
+				}
+			}
+			if len(newv) == 0 {
+				delete(g.waiters, k)
+			} else {
+				g.waiters[k] = newv
+			}
+		}
+	}
+}
+
+func (g *gregorMessageOrderer) WaitForTurn(ctx context.Context, uid gregor1.UID,
+	newVers chat1.InboxVers) (res chan struct{}) {
+	res = make(chan struct{})
+	// Grab latest inbox version if we can
+	vers, err := g.latestInboxVersion(ctx, uid)
+	if err != nil {
+		g.Debug(ctx, "WaitForTurn: failed to get current inbox version: %s", err.Error())
+		close(res)
+		return res
+	}
+
+	// Check for an in-order update
+	if newVers <= vers+1 {
+		close(res)
+		return
+	}
+
+	// Out of order update, we are going to wait a fixed amount of time for the correctly
+	// ordered update
+	deadline := g.clock.Now().Add(time.Second)
+	go func() {
+		g.Lock()
```

Comment:
Is it possible that after getting this lock, the correctly ordered update would be here already?